pilot/pkg/proxy/envoy/v2/cds.go

```diff
@@ -92,3 +99,68 @@ func (s *DiscoveryServer) generateRawClusters(node *model.Proxy, push *model.Pus
 	}
 	return rawClusters, nil
 }
+
+// Set the token path for SDS if SDS_TOKEN_PATH is defined in the proxy metadata
+func SetTokenPathForSdsFromProxyMetadata(c *xdsapi.Cluster, node *model.Proxy) {
+	if sdsTokenPath, found := node.Metadata[model.NodeMetadataSdsTokenPath]; found && len(sdsTokenPath) > 0 {
+		// Set the SDS token path in the TLS certificate config
+		if c.GetTlsContext() != nil && c.GetTlsContext().GetCommonTlsContext() != nil &&
+			c.GetTlsContext().GetCommonTlsContext().GetTlsCertificateSdsSecretConfigs() != nil {
+			for _, sc := range c.GetTlsContext().GetCommonTlsContext().GetTlsCertificateSdsSecretConfigs() {
+				if sc.GetSdsConfig() != nil && sc.GetSdsConfig().GetApiConfigSource() != nil &&
+					sc.GetSdsConfig().GetApiConfigSource().GetGrpcServices() != nil {
+					for _, svc := range sc.GetSdsConfig().GetApiConfigSource().GetGrpcServices() {
+						// If no call-credential in the cluster, no need to set SDS token path
+						if svc.GetGoogleGrpc() != nil && svc.GetGoogleGrpc().GetCallCredentials() != nil &&
+							svc.GetGoogleGrpc().GetCredentialsFactoryName() == model.FileBasedMetadataPlugName {
+							adsLog.Debugf("Set SDS token path in TLS context based on the proxy metadata")
+							svc.GetGoogleGrpc().CallCredentials =
+								model.ConstructgRPCCallCredentials(sdsTokenPath, model.K8sSAJwtTokenHeaderKey)
+						}
+					}
+				}
+			}
+		}
+
+		// Set the SDS token path in the TLS validation context
+		if c.GetTlsContext() != nil && c.GetTlsContext().GetCommonTlsContext() != nil &&
+			c.GetTlsContext().GetCommonTlsContext().GetCombinedValidationContext() != nil &&
+			c.GetTlsContext().GetCommonTlsContext().GetCombinedValidationContext().GetValidationContextSdsSecretConfig() != nil {
+			sc := c.GetTlsContext().GetCommonTlsContext().GetCombinedValidationContext().GetValidationContextSdsSecretConfig()
+			if sc.GetSdsConfig() != nil && sc.GetSdsConfig().GetApiConfigSource() != nil &&
+				sc.GetSdsConfig().GetApiConfigSource().GetGrpcServices() != nil {
+				for _, svc := range sc.GetSdsConfig().GetApiConfigSource().GetGrpcServices() {
+					// If no call-credential in the cluster, no need to set SDS token path
+					if svc.GetGoogleGrpc() != nil && svc.GetGoogleGrpc().GetCallCredentials() != nil &&
+						svc.GetGoogleGrpc().GetCredentialsFactoryName() == model.FileBasedMetadataPlugName {
+						adsLog.Debugf("Set SDS token path in validation context based on the proxy metadata")
+						svc.GetGoogleGrpc().CallCredentials =
+							model.ConstructgRPCCallCredentials(sdsTokenPath, model.K8sSAJwtTokenHeaderKey)
+					}
+				}
+			}
+		}
+	}
+}
+
+func CopyClusters(srcClusters []*xdsapi.Cluster) []*xdsapi.Cluster {
```

Comment:
Please make it start with lowercase.

Reply:
The function name is uppercase because this function and its tests are in different packages.